src/app/application/services/data-source/DataSourcePool.ts

```diff
@@ -0,0 +1,80 @@
+import CallableEventSource, { isClosed } from './CallableEventSource'
+import Router from './Router'
+import SharedPool from './SharedPool'
+
+export type DataSourceResponse<T> = {data: T | undefined, error: Error | undefined, refresh: () => void}
+export type Source = (params: Record<string, unknown>, source: {close: () => void}) => Promise<unknown>
+export type Sources = Record<string, Source>
+
+const create = (src: string, router: Router<Source>): CallableEventSource => {
+  const [path, query] = src.split('?')
+  const queryParams = new URLSearchParams(query)
+  const route = router.match(path)
+  const _source = new CallableEventSource(async function * (this: CallableEventSource) {
+    while (true) {
+      this.readyState = 1
+      // latency
+      // await new Promise(resolve => setTimeout(resolve, 5000))
+      yield route.route({
+        ...{
+          offset: parseInt(queryParams.get('offset') || '0'),
+        },
+        ...route.params,
+      }, this)
+      if (!isClosed(this)) {
+        await new Promise(resolve => setTimeout(resolve, 5000))
+      }
+    }
+  })
+  return _source
+}
+const destroy = (_src: string, source: CallableEventSource) => {
+  if (source) {
+    source.close()
+  }
+}
+
+export class DataSourcePool {
+  cache: Map<string, unknown> = new Map()
+  pool: SharedPool<string, CallableEventSource>
+  constructor(requests: Sources) {
+    const requestRouter: Router<Source> = new Router(requests)
+
+    this.pool = new SharedPool<string, CallableEventSource>(
+      (src: string) => {
+        return create(src, requestRouter)
+      },
+      (src: string, source: CallableEventSource) => {
+        return destroy(src, source)
+      },
+    )
+  }
+
+  source(src: string, ref: symbol) {
+    return this.open(src, ref)
+  }
```

Comment:
Should this method go away if it's just a passthrough to `open` (or inline open in here)?

Reply:
I like this as then you can get an instance of a DataSource as `data` and then call `data.source(...)` on it, so it reads nice, so it's meant to only ever be an alias of `open`. I can remove if you don't like that, but thats exactly what its supposed to be a passthrough and nothing more. lmk

Reply:
Why not remove open then? It's exactly the same function

Reply:
Lemme remove `open`, 2 tics